reflex/state.py

```diff
@@ -1065,26 +1315,126 @@ def get_state(self, token: str) -> State:
             The state for the token.
         """
         if self.redis is not None:
-            redis_state = self.redis.get(token)
+            redis_state = await self.redis.get(token)
             if redis_state is None:
-                self.set_state(token, self.state())
-                return self.get_state(token)
+                await self.set_state(token, self.state())
+                return await self.get_state(token)
             return cloudpickle.loads(redis_state)
 
         if token not in self.states:
             self.states[token] = self.state()
         return self.states[token]
 
-    def set_state(self, token: str, state: State):
+    async def set_state(self, token: str, state: State, lock_id: bytes | None = None):
         """Set the state for a token.
 
         Args:
             token: The token to set the state for.
             state: The state to set.
+            lock_id: If provided, the lock_key must be set to this value to set the state.
+
+        Raises:
+            LockExpiredError: If lock_id is provided and the lock for the token is not held by that ID.
+        """
+        if self.redis is None:
+            return
+        if lock_id is not None:
+            lock_key = f"{token}_lock".encode()
+            # check that we're holding the lock
+            if await self.redis.get(lock_key) != lock_id:
+                raise LockExpiredError(
+                    f"Lock expired for token {token} while processing. Consider increasing "
+                    f"`app.state_manager.lock_expiration` (currently {self.lock_expiration}) "
+                    "or use `@rx.background` decorator for long-running tasks."
+                )
+        await self.redis.set(token, cloudpickle.dumps(state), ex=self.token_expiration)
+
+    @contextlib.asynccontextmanager
+    async def _redis_lock(self, token: str):
+        """Obtain a redis lock for a token.
+
+        Args:
+            token: The token to obtain a lock for.
+
+        Yields:
+            The ID of the lock (to be passed to set_state).
+
+        Raises:
+            RuntimeError: If redis is not configured.
+            LockExpiredError: If the lock has expired while processing the event.
         """
         if self.redis is None:
+            raise RuntimeError("Redis is not configured")
+        lock_key = f"{token}_lock".encode()
+        lock_id = uuid.uuid4().hex.encode()
+        lock_key_channel = f"__keyspace@0__:{lock_key.decode()}"
+
+        def try_get_lock():
```

Comment:
Might be good to define this outside of the function as it will get redefined every time this is called.